### ISSUE TITLE
Merging bug if there is the same property/value between them

### ISSUE THREAD

**Issue by jepek** [jakubpawlowicz/clean-css#773]
I have found a bug with merging. What I noted is that the bug occurs when there is the same property/value between duplicated selectors. For example, this works fine:

```
#test {
    background: red;
}

#test {
    height: 7.917em;
}

#test {
    background-color: pink;
}

```

the output is:

```
#test{background:pink;height:7.917em}
```

But if there is some other selector with the same property/value declaration:

```
#test {
    background: red;
}

.other-selector {
    background-color: #f9f9f9;
}

#test {
    height: 7.917em;
}

#test {
    background-color: pink;
}

```

we are in trouble

```
#test{background:red}.other-selector{background-color:#f9f9f9}#test{height:7.917em;background-color:pink}
```

**Comment by jakubpawlowicz** [comment]
clean-css still provides a valid result but in this very case `#test` has a higher specificity than `.other-selector` so reordering is fine - adding it as enhancement.

**Comment by jakubpawlowicz** [comment]
I've just pushed `773-specificity` branch with this feature. We apparently missed reordering rules based on selector specificity which are ready now.

After some testing I made an assumption that two selectors can be reordered if their specificity is different or if it's the same but there are no property clashes. I hope this is right but it'd be great to do some more testing before 4.0 lands.

The second commit about speeding up specificity lookup is about caching specificity calculations as it seems to be an expensive step.
